src/netops.c

```diff
@@ -280,3 +280,89 @@ int gitno_extract_url_parts(
 
 	return 0;
 }
+
+int gitno_extract_url_data_parts(
+	char **scheme,
+	char **host,
+	char **port,
+	char **path,
+	char **username,
+	char **password,
+	char **query,
+	const char *url)
+{
+	struct http_parser_url u = { 0 };
+	const char *_userinfo;
+
+	if(http_parser_parse_url(url, strlen(url), false, &u)) {
+		giterr_set(GITERR_NET, "Malformed URL '%s'", url);
+		return GIT_EINVALIDSPEC;
+	}
+
+#define GITNO_COPY_URL_PART(dest, field) \
```

Comment:
The code here needs to be in a `do { } while (0)` block to guard against uses with optional braces.